index.bs

```diff
@@ -1462,3 +1484,231 @@ The [=remote end steps=] with |command parameters| are:
+       |command parameters| if present, or null otherwise.
+
+  1. Let |parent| be the result of [=trying=] to [=get a browsing context=]
+     given |parent id|.
+
+  1. If |parent| is null, let |contexts| be a list containing all [=top-level
+     browsing contexts=]. Otherwise let |contexts| be a list containing all
+     [=/browsing contexts=] which are [=child browsing contexts=] of |parent|.
+
+  1. Let |context info| be a list.
+
+  1. For each |context| of |context|, append the result of [=get the browsing
+     context info=] given |context| to |context info|.
+
+  1. Let |body| be a map matching the <code>BrowsingContextListResult</code>
+     production, with the <code>contexts</code> field set to |context info|.
+
+  1. Return [=success=] with data |body|.
+
+Issue: Should this contain other data like the current URL for the context.
+
+</div>
+
+### Events ### {#module-contexts-events}
+
+#### contextCreated #### {#event-browsingContext-contextCreated}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        BrowsingContextContextCreatedEvent = {
+         method: "browsingContext.contextCreated",
+         params: BrowsingContextInfo
+       }
+      </pre>
+   </dd>
+</dl>
+The [=remote end event trigger=] is:
+
+<div algorithm="remote end event trigger for browsingContext.contextCreated">
+When the [=create a new browsing context=] algorithm is invoked, after the
+[=active document=] of the browsing context is set, run the following steps:
+
+ 1. Let |context| be the newly created browsing context.
+
+ 1. Let |related contexts| be a set containing the [=parent browsing context=]
+    of |context|, if that is not null, or an empty set otherwise.
+
+ 1. Let |params| be the result of [=get the browsing context info=] given
+    |context|.
+
+ 1. Let |body| be a map matching the
+    <code>BrowsingContextContextCreatedEvent</code> production, with the
+    <code>value</code> field set to |params|.
+
+ 1. [=Emit an event=] with |body| and |related contexts|.
+
+</div>
+
+#### contextDestroyed #### {#event-browsingContext-contextDestroyed}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        BrowsingContextContextDestroyedEvent = {
+         method: "browsingContext.contextDestroyed",
+         params: BrowsingContextDestroyedParameters
+       }
+
+       BrowsingContextDestroyedParameters = {
+         context: BrowsingContext,
+         parent: BrowsingContext / null,
+       }
+      </pre>
+   </dd>
+</dl>
+The [=remote end event trigger=] is:
+
+<div algorithm="remote end event trigger for browsingContext.contexDestroyed">
+
+When [=a browsing context is discarded=], run the following steps:
+
+ 1. Let |context| be the browsing context being discarded.
+
+ 1. Let |context id| be the [=browsing context id=] for |context|.
+
+ 1. If |context| has a [=parent browsing context=] let |parent| be that parent
```

Comment:
Have you checked how this will end up working if a whole tree of iframes are removed at once? Will events for each browsing context be fired, and in what order?

Reply:
Hmm, so per spec I think the answer is "yes" and "currently top-down". But it depends on the exact details of when the algorithm is invoked. Step 1 discards all documents in the session history which also discards child browsing contexts of those documents. So I guess we actually want to only emit events for contexts that have an active document. And probably want the events to happen bottom up, so that child contexts emit the event before their parent? That seems like it would give stronger invariants.

Reply:
I've changed this a bit, but it remains very fragile in the way it integrates with HTML. I feel like we're going to end up revisting this later.

Reply:
I filed it before I got to this comment, but https://github.com/whatwg/html/issues/6194 might be where we figure this out.

Here's my current thinking. Rather than talking about context creation and destruction of individual browsing contexts, express this in terms of modifications to the browsing context tree. Creation/insertion is still pretty simple, I think a newly inserted browsing context can't yet have any descendant browsing siblings. But in the case of removal, send along the whole frame tree that was just removed in a single event.